File: Ruby/1-6-3.rb
# -*- coding: utf-8 -*-
# PCCB 1-6-3

class Solver
	def self.solve(s, a)
    prepare!(a)
    for i in 0...a.size
      for j in (i+1)...a.size
        for k in (j+1)...a.size
          l = binary_search(a, k, a.size - 1, s - a[i] - a[j] - a[k])
          if l > -1
            puts "answer: #{i} #{j} #{k} #{l}"
            return
          end
        end
      end
    end
	end

  def self.prepare!(a)
    a = a.kind_of?(Array) && a.all? {|x| x.is_a?(Integer)} ? a : []
    a.sort! {|x, y| x <=> y}
  end

  def self.binary_search(a, l, r, t)
    if l == r
      return a[l] == t ? l : -1
    end
    m = (l + r) / 2
    if a[m] < t
      return binary_search(a, m + 1, r, t)
    else
      return binary_search(a, l, m, t)
    end
  end
end

$sum   = 0
$array = []

ARGV.each_with_index do |arg, index|
	begin
    if index == 0
      $sum    = Integer(arg)
    else
      $array << Integer(arg)
    end
	rescue Exception => e
    puts e.to_s
	end
end

Solver.solve($sum, $array)

File: Python/1-6-3.py
# PCCB 1-6-3
# -*- coding: utf-8 -*-

import sys

class Solver(object):
    @staticmethod
    def solve(target, items):
        Solver.prepare(items)
        for i in range(0, len(items)):
            for j in range(i + 1, len(items)):
                for k in range(j + 1, len(items)):
                    l = Solver.binary_search(target - items[i] - items[j] - items[k], items[k + 1:])
                    if l >= 0:
                        return items[i], items[j], items[k], items[l + k + 1]
        return None

    @staticmethod
    def prepare(items):
        items = sorted(items)

    @staticmethod
    def binary_search(target, items):
        left = 0
        right = len(items) - 1
        while left < right:
            mid = (left + right) // 2
            if target <= items[mid]:
                right = mid
            else:
                left = mid + 1
        return left if target == items[left] else -1


def main(argv=None):
    if not argv:
        argv = sys.argv[1:]
    if argv:
        try:
            target = int(argv.pop(0))
            items = [int(arg) for arg in argv]
            candidate = Solver.solve(target, items)
            if candidate:
                print "found: %d %d %d %d" % candidate
            else:
                print "not found"
        except Exception as e:
            print >> sys.stderr, str(e)
            exit
    else:
        print >> sys.stderr, "usage: python 1-6-3.py <target> <items>"


if __name__ == '__main__':
    sys.exit(main())
